### PR TITLE
improve accuracy of logsoftmax computation on cuda

### LINKED ISSUE #38839: CUDA log_softmax produce wrong result for large inputs
```
In [12]: import torch
    ...:
    ...: a = torch.rand(4,3)
    ...: b = a - 1e7
    ...: c = torch.nn.functional.log_softmax(b, 1)
    ...: print(c)
    ...:
    ...: b = b.cuda()
    ...: c = torch.nn.functional.log_softmax(b, 1)
    ...: print(c)
tensor([[-1.0986, -1.0986, -1.0986],
        [-1.5514, -1.5514, -0.5514],
        [-0.8620, -0.8620, -1.8620],
        [-1.8620, -0.8620, -0.8620]])
tensor([[-1., -1., -1.],
        [-2., -2., -1.],
        [-1., -1., -2.],
        [-2., -1., -1.]], device='cuda:0')
```
CPU result is correct. 
reported by @hengyuan-hu

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by ngimel**
Fixes #38839. Previously, if magnitude of input values was large, when computing `max+log(sum)` the `log(sum)` value was essentially ignored, now the result is computed as
`x-max-log(sum)` which has a better chance of preserving accuracy.

### CODE REVIEW

**Review by mruberry — APPROVED** [review]
Awesome! That's cool you found the issue with test_ctc_loss, too!

**Review by albanD — APPROVED** [review]
LGTM
